rkyv/src/niche/niched_option.rs

```diff
@@ -1,81 +1,35 @@
-//! A niched archived `Option<T>` that uses less space based on a [`Niching`].
+//! A niched `ArchivedOption<T>` that uses less space based on a [`Niching`].
 
-use core::{cmp, fmt, mem::ManuallyDrop, ops::Deref};
+use core::{cmp, fmt, marker::PhantomData, mem::MaybeUninit, ops::Deref};
 
 use munge::munge;
 use rancor::Fallible;
 
 use super::niching::Niching;
-use crate::{seal::Seal, Archive, Archived, Place, Portable, Serialize};
+use crate::{seal::Seal, Archive, Place, Portable, Serialize};
 
-/// A niched archived `Option<T>`.
+/// A niched `ArchivedOption<T>`.
 ///
 /// It uses less space by storing the `None` variant in a custom way based on
 /// `N`.
-#[derive(Portable)]
-#[rkyv(crate)]
-#[cfg_attr(feature = "bytecheck", derive(bytecheck::CheckBytes))]
 #[repr(transparent)]
-pub struct NichedOption<T, N>
-where
-    T: Archive,
-    N: Niching<T::Archived> + ?Sized,
-{
-    repr: Repr<T::Archived, N>,
-}
-
-#[repr(C)]
-#[derive(Portable)]
-#[rkyv(crate)]
-union Repr<T, N: Niching<T> + ?Sized> {
-    some: ManuallyDrop<T>,
-    niche: ManuallyDrop<<N as Niching<T>>::Niched>,
+pub struct NichedOption<T, N: ?Sized> {
+    repr: MaybeUninit<T>,
+    _niching: PhantomData<N>,
 }
 
-impl<T, N: Niching<T> + ?Sized> Repr<T, N> {
-    /// Compile-time check to make sure that the niched type is not greater than
-    /// the archived type.
-    ///
-    /// ```compile_fail
-    /// use rkyv::{
-    ///     niche::{niching::Niching, niched_option::NichedOption},
-    ///     Archived, Place,
-    /// };
-    ///
-    /// type T = u16;
-    /// type N = u32;
-    ///
-    /// struct UselessNiching;
-    ///
-    /// unsafe impl Niching<Archived<T>> for UselessNiching {
-    ///     type Niched = Archived<N>;
-    ///
-    ///     fn is_niched(_: &Archived<N>) -> bool {
-    ///         false
-    ///     }
-    ///
-    ///     fn resolve_niched(_: Place<Self::Niched>) {}
-    /// }
-    ///
-    /// let archived: Archived<N> = 456.into();
-    /// let niched: &NichedOption<T, UselessNiching> =
-    ///     unsafe { std::mem::transmute(&archived) };
-    /// let _ = niched.is_none(); // <- size check = compile error
-    /// ```
-    const NICHE_SIZE_CHECK: () = {
-        if size_of::<<N as Niching<T>>::Niched>() > size_of::<T>() {
-            panic!(
-                "`N::Niched` is greater than `T` and thus useless for niching"
-            );
-        }
-    };
-}
+// SAFETY: The safety invariant of `Niching<T>` requires its
+// implementor to ensure that the contained `MaybeUninit<T>` is
+// portable and thus implies this safety.
+unsafe impl<T, N: Niching<T> + ?Sized> Portable for NichedOption<T, N> {}
```

Comment:
`MaybeUninit` does not implement `Portable` so I don't think `NichedOption` can derive it. The `T: Portable` bound is something I overlooked and should indeed be added though.

Arguably, even `<N as Niching<T>>::Niched: Portable` should be added since that type will override some portion of the `T` but I think it's fine to omit that and instead leverage the `Niching` safety invariant. That way, `Niching` implementors can use a non-portable `Niched` type for which they have to guarantee themselves that the way it is used will always be portable.

Reply:
> MaybeUninit does not implement Portable so I don't think NichedOption can derive it.

I think it's safe for us to address that issue by just implementing `Portable` for `MaybeUninit`. I made that change in bb1c04a54f511f63e5be5ba042e82584141cd3c6.